Replace pydantic with mashumaro

bvanelli opened the pull request:
# Proposed Changes

This PR exchanges the pydantic dependency with [mashumaro](https://github.com/Fatal1ty/mashumaro). This is due to the fact that the version of pydantic used is not anymore supported, and a migration was not accepted (see https://github.com/frenck/python-demetriek/pull/531)

I've taken into account the following PRs from the author as references on how to do the migration:

- https://github.com/frenck/python-elgato/pull/938
- https://github.com/frenck/python-tailscale/pull/443

# Changelog

I had to to some unrelated changes to get the pipeline to run through, due to some library upgrades that now fail on CI. Since the content is not crazy big, I decided to add them to the same PR, but let me know if I should make a separate PR instead. Here is the list of changes:

- Exchanged pydantic to mashumaro, like it was done on the other repositories

# Discussion

As far as tests are concerned, the changes should work out of the box, but I do not have one of those devices to test it. If there is a way to have simulated APIs or there are test cases that should be covered, let me know.

# Related Issues

For the pydantic upgrade in HA, this change is necessary:

- https://github.com/home-assistant/core/issues/99218

allenporter commented:
> # Changelog
> I had to to some unrelated changes to get the pipeline to run through, due to some library upgrades that now fail on CI. 
> Since the content is not crazy big, I decided to add them to the same PR, but let me know if I should make a separate PR instead. 

I'd definitely say make each of these changes into separate PRs for each change mentioned (CI fixes, async timeout, grammar/spelling fixes,  linting fixes, etc)

bvanelli commented:
I moved the unrelated changes to get the pipelines to run to other PRs:

- File had to be reformatted with latest ruff for the CI workflow to succeed, so some styling changed here and there, like spacings between and within comments.: https://github.com/frenck/python-demetriek/pull/742
- On the file `.github/CODE_OF_CONDUCT.md`, the codespell pre-commit was failing for the word `socio-economic`, so I changed it to `socioeconomic`. Seems like the [word is not hyphenated](https://www.collinsdictionary.com/de/worterbuch/englisch/socioeconomic): https://github.com/frenck/python-demetriek/pull/743
- On CI linting workflow, the command `poetry run ruff .` was changed to `poetry run ruff check .`: https://github.com/frenck/python-demetriek/pull/741
- I had to adapt the upload of artifacts because of https://github.com/actions/upload-artifact/issues/602, since the `.coverage` file would not be uploaded anymore without the `include-hidden-files` option.: https://github.com/frenck/python-demetriek/pull/744

Those can be made first and I'll rebase this branch once that is done.

frenck reviewed (changes requested):
Thanks, for your work on this @bvanelli 👍 

I've merged all split out PRs, and went over the project in certain places myself as well. CI on the main branch now passes again.

Can you rebase this PR on top of the latest `main` branch?

Thanks! 👍 

../Frenck

bvanelli commented:
@frenck did the rebase, everything looks nice except the poetry file that has some references to my own environent after running `poetry lock --no-update` to remove the dependency. You can also replace it on merge.

Let me know if I should change something else 👍

frenck commented on src/demetriek/models.py:
```diff
@@ -50,41 +56,48 @@ class Bluetooth(BaseModel):
     height: int
-    display_type: DisplayType | None = Field(default=None, alias="type")
     screensaver: DisplayScreensaver
+    display_type: DisplayType | None = field(
```
We can avoid moving this stuff around by setting `kw_only=True` on the dataclasses.

frenck reviewed (changes requested):
There seems to be a serialization issue that is still left, the example provided in this repository doesn't work right now:

![CleanShot 2024-12-09 at 17 04 40@2x](https://github.com/user-attachments/assets/024c2f6b-63c2-4363-8737-4771c67fe1b7)

frenck commented:
With the last change in place, it would only omit `None` values from the top level notification object, and not its contents, right?

bvanelli commented:
> With the last change in place, it would only omit `None` values from the top level notification object, and not its contents, right?

Yes, that is correct, I did that based on the [docs request](https://lametric-documentation.readthedocs.io/en/latest/reference-docs/device-notifications.html#body), and there it does not seem clear to me that those sub-fields are optional, for example, for `Model.sound`:

> object that describes the notification sound to play when notification pops on the LaMetric Time’s screen

Fields from the `Notification` have the clear *Optional* keyword, but not this one, so I guess they are not optional on the request.

I did not move to review again since there is another issue. It seems like `__pre_deserialize__` does not work the same way as the root validator does, since it does only run when desserializing, but not when directly instancing the object. I'm currently looking for a way to do that.

bvanelli commented:
> I'm currently looking for a way to do that.

Never mind, I just realized I used the wrong validation, I should have used dataclasses `__post_init__` instead.

bvanelli commented:
@frenck Thanks for reviewing and noticing my mistake on the serialization. I have changed the `__pre_deserialize__` to `__post_init__` since `Sound` is only instantiated directly.

Regarding the omit optional fields, I have not changed the behaviour for the inner objects. Since you have more experience with the API, should that be the case? That is affectin:

- `Model.sound`
- `Goal.icon`
- `GoalData.icon`
- `Simple.icon`

and from what I understood from the documentation none of them are option.

frenck commented:
Awesome 👍 

Just tested it against real devices, it seems to work as expected 👍 

Will push up some nitpick sorting we can do now and continue on merging this one in.

../Frenck
